yax/core/src/main/scala/doobie/util/log.scala

```diff
@@ -0,0 +1,109 @@
+package doobie.util
+
+import doobie.imports._
+#+scalaz
+import scalaz._, Scalaz._
+#-scalaz
+#+cats
+import cats._, cats.data._, cats.functor.Contravariant
+#-cats
+import java.util.logging.{ Logger, Level }
+import scala.Predef.augmentString
+import shapeless.HList
+import shapeless.ops.hlist.ToTraversable
+
+/** A module of types and instances for logged statements. */
+object log {
+  import LogEvent._
+
+  sealed abstract class LogEvent[A] extends Product with Serializable {
+
+    /** The complete SQL string as seen by JDBC. */
+    def sql: String
+
+    /** The query arguments. This is usually an HList. */
+    def args: A
+
+    /** LogEvent is a covariant functor. */
+    def map[B](f: A => B): LogEvent[B] =
+      this match {
+        case Success (s, a, m1, m2)         => Success(s, f(a), m1, m2)
+        case ProcessingFailure(s, a, m1, t) => ProcessingFailure(s, f(a), m1, t)
+        case ExecFailure(s, a, t)           => ExecFailure(s, f(a), t)
+      }
+
+  }
+  object LogEvent {
+
+    case class Success          [A](sql: String, args: A, msExec: Long, msProcessing: Long) extends LogEvent[A]
+    case class ProcessingFailure[A](sql: String, args: A, msExec: Long, failure: Throwable) extends LogEvent[A]
```

Comment:
I wouldn't mind a comment on this class explaining that it indicates failure in the user's code when processing the result of a query. Initially I assumed that this was an error coming out of JDBC when preparing the query.